### PR TITLE
Add `ASTReduce` variant that provides key and value for dictionaries

### PR DESCRIPTION
**Pull request by monkeydom**
Also add `ASTMapPairs` to map from dictionary to array

### CODE REVIEW

File: Specs/ASTMapSpec.m [hunk 1]
```diff
@@ -98,6 +102,24 @@ - (void)testEverything {
...
+
+            NSArray *after = ASTMapPairs(before, ^(id key, id value) {
+                return keyValue(key, value);
+            });
+            after = [after sortedArrayUsingSelector:@selector(compare:)];
+
+            XCTAssertEqualObjects(after, (@[
+                @"en: Hello",
+                @"fr: Bonjour",
+                                          ]));
```
Comment (robb) [inline Specs/ASTMapSpec.m:119]: WS
```suggestion
            ]));
```

File: Specs/ASTMapSpec.m [hunk 2]
```diff
@@ -98,6 +102,24 @@ - (void)testEverything {
                                           }));
```
Comment (robb) [inline Specs/ASTMapSpec.m:102]:
```suggestion
	            }));
```

File: Asterism/ASTMap.m [hunk 3]
```diff
@@ -68,6 +69,32 @@
...
+//
+//    ASTEach(dict, ^(id key, id obj) {
+//        id transformed = block(key, obj);
+//
+//        if (transformed != nil) {
+//            [result addObject:transformed];
+//        }
+//    });
+//
+//    return result;
```
Comment (robb) [inline Asterism/ASTMap.m:94]: I think this can be removed?

```suggestion
```

File: Specs/ASTReduceSpec.m [hunk 4]
```diff
@@ -43,6 +47,17 @@ - (void)testEverything {
 
                     XCTAssertEqualObjects(ASTReduce(dict, @4, add), @10);
                 }];
+                
+                [XCTContext runActivityNamed:@"should be able to reduce key value pairs in dictionary" block:^(id<XCTActivity> _Nonnull activity){
```
Comment (robb) [inline Specs/ASTReduceSpec.m:51]: Can you move this out of the `for objects implementing <NSFastEnumeration>` scope?

File: Specs/ASTMapSpec.m [hunk 5]
```diff
@@ -98,6 +102,24 @@ - (void)testEverything {
...
         }];
 
+        [XCTContext runActivityNamed:@"should return a dictioanry" block:^(id<XCTActivity> _Nonnull activity){
```
Comment (robb) [inline Specs/ASTMapSpec.m:105]: Can you move this out of the `for arrays` scope? Also typo in `dictioanry`.

**Review by robb — APPROVED** [review]
Looks good, just a few minor notes ✨

**Comment by robb** [comment]
Also, can you merge in `master` which should get GitHub to run CI?

File: Specs/ASTMapSpec.m (hunk 5 above)
Comment (monkeydom) [reply Specs/ASTMapSpec.m:105]: Actually the comment should be an array, and the target is a dictionary, so not sure if for arrays or for dictionaries is the right place here. You sure about the for arrays scope?

File: Specs/ASTMapSpec.m [hunk 6]
```diff
@@ -95,9 +99,27 @@ - (void)testEverything {
             XCTAssertEqualObjects(after, (@{
                 @"fr": @"BONJOUR",
                 @"en": @"HELLO"
-                                          }));
+            }));
         }];
 
+        [XCTContext runActivityNamed:@"should return an array" block:^(id<XCTActivity> _Nonnull activity){
```
Comment (robb) [inline Specs/ASTMapSpec.m:105]: I guess this shouldn't go into the `for arrays` scope? Either dictionary or its own is fine I guess.

**Review by robb — APPROVED** [review]
Just one note.

File: Specs/ASTMapSpec.m (hunk 6 above)
Comment (monkeydom) [reply Specs/ASTMapSpec.m:105]: It is in the for `for dictionaries` scope currently. Am I missing something?

Comment (robb) [reply Specs/ASTMapSpec.m:105]: My bad!

**Review by robb — APPROVED** [review]

**Comment by robb** [comment]
Thanks again

**Pull request merged**